File: proto/scouting.py
"""
Project Dugout - Scouting Engine

이 모듈의 목적은 **플레이어와 AI에게 틀린 정보를 주는 것**이다.

원칙
  1. 팀은 진실이 아니라 자기 스카우트의 추정치로 판단한다.
  2. 오차는 (팀, 선수) 쌍마다 **고정**된다. 같은 선수를 다시 봐도 추정이
     흔들리지 않는다. 관찰을 늘리면 오차의 '크기'만 줄고 방향은 유지된다.
     → 한 팀이 특정 선수를 계속 과대평가하는 일이 자연스럽게 생긴다.
  3. 잠재력 오차가 현재 능력 오차보다 훨씬 크다. 이게 드래프트의 재미다.
  4. **오차의 일부는 리그 전체가 공유한다.** 팀마다 독립적으로만 틀리면
     드래프트 순위가 개별 스카우트보다 훨씬 정확해져 버린다 (오차가 평균되므로).
     실제 드래프트에서는 모든 구단이 같은 정보를 보고 같은 방향으로 틀린다.
     그래서 전체 1순위가 통째로 망하는 일이 벌어진다.
"""
import math
import development as dev
from pa_engine import Batter

# 기본 오차 (20-80 스케일, 1 표준편차)
SIGMA_CUR_AMATEUR = 6.5
SIGMA_POT_AMATEUR = 16.0
SIGMA_CUR_PRO = 2.2          # 1군 기록이 쌓인 선수는 현재 능력을 거의 안 틀린다
SIGMA_POT_PRO = 7.0

# 오차 중 리그 전체가 공유하는 비율 (분산 기준)
CONSENSUS_SHARE = 0.72


class ScoutingDept:
    """팀의 스카우트 조직. 개별 스카우트 대신 부서 단위로 추상화했다."""

    def __init__(self, rng):
        self.eval_current = max(20.0, min(80.0, rng.gauss(50, 11)))
        self.eval_potential = max(20.0, min(80.0, rng.gauss(50, 12)))
        self.hitting = max(20.0, min(80.0, rng.gauss(50, 10)))
        self.pitching = max(20.0, min(80.0, rng.gauss(50, 10)))
        # 구단 성향 편향: 이 팀이 과대평가하는 툴 (스카우트의 취향)
        self.bias = {
            "speed": rng.gauss(0, 2.6), "hr_power": rng.gauss(0, 2.6),
            "stuff": rng.gauss(0, 2.6), "contact": rng.gauss(0, 1.8),
            "command": rng.gauss(0, 1.8), "discipline": rng.gauss(0, 2.2),
        }
        self.memory = {}     # pid -> {attr: 표준정규 오차 시드}
        self.looks = {}      # pid -> 관찰 횟수

    def _seed(self, p, rng):
        if not hasattr(p, "scout_consensus"):      # 리그 공통 오차 (선수당 한 번)
            p.scout_consensus = {a: rng.gauss(0, 1) for a in dev.attrs_of(p)}
            p.scout_consensus_pot = {a: rng.gauss(0, 1) for a in dev.attrs_of(p)}
        if p.pid not in self.memory:
            self.memory[p.pid] = {a: rng.gauss(0, 1) for a in dev.attrs_of(p)}
            self.memory[p.pid]["_pot"] = {a: rng.gauss(0, 1) for a in dev.attrs_of(p)}
        return self.memory[p.pid]

    def _quality(self, p, which):
        """스카우트 능력 → 오차 배율. 좋은 스카우트일수록 작다."""
        base = self.eval_current if which == "cur" else self.eval_potential
        know = self.hitting if isinstance(p, Batter) else self.pitching
        z = ((base - 50) / 10) * 0.75 + ((know - 50) / 10) * 0.25
        return max(0.45, 1.30 - 0.32 * z)

    def observe(self, p, rng, n=1):
        self._seed(p, rng)
        self.looks[p.pid] = self.looks.get(p.pid, 0) + n

    def report(self, p, rng, is_pro=False):
        """스카우팅 리포트. (능력별 추정 구간, 추정 OVR 구간, 추정 잠재력 구간)"""
        seed = self._seed(p, rng)
        looks = self.looks.get(p.pid, 0)
        shrink = 1.0 / math.sqrt(1.0 + 0.40 * looks)
        sc = (SIGMA_CUR_PRO if is_pro else SIGMA_CUR_AMATEUR) * shrink * self._quality(p, "cur")
        sp = ((SIGMA_POT_PRO if is_pro else SIGMA_POT_AMATEUR) * shrink
              * self._quality(p, "pot") * getattr(p, "scout_difficulty", 1.0))

        kc = CONSENSUS_SHARE ** 0.5            # 공유 오차 가중
        ki = (1.0 - CONSENSUS_SHARE) ** 0.5    # 팀 고유 오차 가중
        est_cur, est_pot = {}, {}
        for a in dev.attrs_of(p):
            b = self.bias.get(a, 0.0)
            ec = kc * p.scout_consensus[a] + ki * seed[a]
            ep = kc * p.scout_consensus_pot[a] + ki * seed["_pot"][a]
            est_cur[a] = _clamp(getattr(p, a) + ec * sc + b)
            # 잠재력 추정은 현재 능력 추정 아래로 내려가지 않는다
            v = p.pot[a] + ep * sp + b * 1.3
            est_pot[a] = _clamp(max(v, est_cur[a]))

        return Report(p, est_cur, est_pot, sc, sp, looks)


def _clamp(v):
    return max(20.0, min(80.0, v))


class Report:
    def __init__(self, p, est_cur, est_pot, sc, sp, looks):
        self.p, self.est_cur, self.est_pot = p, est_cur, est_pot
        self.sigma_cur, self.sigma_pot, self.looks = sc, sp, looks

    def _ovr(self, table):
        saved = {a: getattr(self.p, a) for a in table}
        for a, v in table.items():
            setattr(self.p, a, v)
        o = dev.overall(self.p)
        for a, v in saved.items():
            setattr(self.p, a, v)
        return o

    @property
    def ovr(self):
        return self._ovr(self.est_cur)

    @property
    def pot(self):
        return self._ovr(self.est_pot)

    @property
    def confidence(self):
        """0~100. 관찰량과 스카우트 능력이 만드는 확신도."""
        return max(5.0, min(99.0, 100.0 * (1.0 - self.sigma_pot / 18.0)))

    def range_of(self, attr, which="cur"):
        """플레이어에게 보여줄 구간. '컨택 53~63' 형태."""
        v = (self.est_cur if which == "cur" else self.est_pot)[attr]
        s = (self.sigma_cur if which == "cur" else self.sigma_pot) * 0.9
        return _clamp(v - s), _clamp(v + s)

    def ovr_range(self, which="cur"):
        v = self.ovr if which == "cur" else self.pot
        s = (self.sigma_cur if which == "cur" else self.sigma_pot) * 0.50
        return _clamp(v - s), _clamp(v + s)

    def text(self):
        """스카우팅 코멘트. 숨은 성향은 숫자가 아니라 문장으로만 노출한다."""
        h = self.p.hidden
        out = []
        we = h["work_ethic"]
        out.append("훈련 태도가 대단히 성실하다." if we >= 65 else
                   "성실한 선수로 평가된다." if we >= 52 else
                   "훈련 태도에 관한 우려가 있다." if we < 38 else
                   "훈련 태도는 평범하다.")
        pr = h["professionalism"]
        if pr >= 65: out.append("자기 관리가 뛰어나다.")
        elif pr < 35: out.append("프로 의식에 물음표가 붙는다.")
        ip = h["injury_prone"]
        if ip >= 63: out.append("부상 이력과 체질이 걱정스럽다.")
        elif ip <= 37: out.append("몸이 튼튼하다.")
        if self.looks <= 1:
            out.append("관찰 기회가 적어 평가의 불확실성이 크다.")
        elif self.looks >= 3:
            out.append("장기간 추적 관찰한 선수다.")
        return " ".join(out)

File: web/js/core/scouting.js
// 스카우팅: 팀에게 틀린 정보를 준다. 오차의 72%는 리그 전체가 공유한다.
import * as dev from './development.js';

export const SIGMA_CUR_AMATEUR = 6.5, SIGMA_POT_AMATEUR = 16.0;
export const SIGMA_CUR_PRO = 2.2, SIGMA_POT_PRO = 7.0;
export const CONSENSUS_SHARE = 0.72;
export const FOREIGN_OPTIMISM = 3.4;
// 외국인은 잠재력뿐 아니라 '지금 이 리그에서 얼마나 하느냐'가 불확실하다.
// 국내 프로는 여기서 뛴 기록이 있지만, 외국인은 번역해야 한다.
export const FOREIGN_CUR_MULT = 1.95;

const clamp = (v) => Math.max(20, Math.min(80, v));

export class ScoutingDept {
  constructor(rng) {
    this.eval_current = clamp(rng.gauss(50,11));
    this.eval_potential = clamp(rng.gauss(50,12));
    this.hitting = clamp(rng.gauss(50,10));
    this.pitching = clamp(rng.gauss(50,10));
    this.bias = { speed: rng.gauss(0,2.6), hr_power: rng.gauss(0,2.6),
      stuff: rng.gauss(0,2.6), contact: rng.gauss(0,1.8),
      command: rng.gauss(0,1.8), discipline: rng.gauss(0,2.2) };
    this.memory = new Map();
    this.looks = new Map();
  }
  _seed(p, rng) {
    if (!p.scout_consensus) {
      p.scout_consensus = {}; p.scout_consensus_pot = {};
      for (const a of dev.attrsOf(p)) { p.scout_consensus[a] = rng.gauss(0,1);
                                        p.scout_consensus_pot[a] = rng.gauss(0,1); }
    }
    if (!this.memory.has(p.pid)) {
      const cur = {}, pot = {};
      for (const a of dev.attrsOf(p)) { cur[a] = rng.gauss(0,1); pot[a] = rng.gauss(0,1); }
      this.memory.set(p.pid, { cur, pot });
    }
    return this.memory.get(p.pid);
  }
  _quality(p, which) {
    const base = which === 'cur' ? this.eval_current : this.eval_potential;
    const know = p.kind === 'B' ? this.hitting : this.pitching;
    const z = ((base-50)/10)*0.75 + ((know-50)/10)*0.25;
    return Math.max(0.45, 1.30 - 0.32*z);
  }
  observe(p, rng, n = 1) { this._seed(p, rng); this.looks.set(p.pid, (this.looks.get(p.pid) ?? 0) + n); }
  report(p, rng, isPro = false, staffMult = 1) {
    const seed = this._seed(p, rng);
    const looks = this.looks.get(p.pid) ?? 0;
    const shrink = staffMult / Math.sqrt(1 + 0.40*looks);
    const sc = (isPro ? SIGMA_CUR_PRO : SIGMA_CUR_AMATEUR) * shrink * this._quality(p,'cur')
      * (p.foreign ? FOREIGN_CUR_MULT : 1);
    const sp = (isPro ? SIGMA_POT_PRO : SIGMA_POT_AMATEUR) * shrink * this._quality(p,'pot')
      * (p.scout_difficulty ?? 1.0);
    const kc = Math.sqrt(CONSENSUS_SHARE), ki = Math.sqrt(1 - CONSENSUS_SHARE);
    // 외국인은 다른 리그 성적으로 판단한다. 그 성적은 이쪽으로 그대로 옮겨오지
    // 않는데, 보고서는 그걸 충분히 깎지 못한다. 리그 전체가 같은 방향으로 틀린다.
    const fb = p.foreign ? FOREIGN_OPTIMISM : 0;
    const estCur = {}, estPot = {};
    for (const a of dev.attrsOf(p)) {
      const b = this.bias[a] ?? 0;
      const ec = kc*p.scout_consensus[a] + ki*seed.cur[a];
      const ep = kc*p.scout_consensus_pot[a] + ki*seed.pot[a];
      estCur[a] = clamp(p[a] + ec*sc + b + fb);
      estPot[a] = clamp(Math.max(p.pot[a] + ep*sp + b*1.3 + fb, estCur[a]));
    }
    return new Report(p, estCur, estPot, sc, sp, looks);
  }
}

export class Report {
  constructor(p, estCur, estPot, sc, sp, looks) {
    this.p = p; this.estCur = estCur; this.estPot = estPot;
    this.sigmaCur = sc; this.sigmaPot = sp; this.looks = looks;
  }
  _ovr(table) {
    const saved = {};
    for (const a in table) { saved[a] = this.p[a]; this.p[a] = table[a]; }
    const o = dev.overall(this.p);
    for (const a in saved) this.p[a] = saved[a];
    return o;
  }
  get ovr() { return this._ovr(this.estCur); }
  get pot() { return this._ovr(this.estPot); }
  get confidence() { return Math.max(5, Math.min(99, 100*(1 - this.sigmaPot/18))); }
  rangeOf(attr, which='cur') {
    const v = (which==='cur'?this.estCur:this.estPot)[attr];
    const s = (which==='cur'?this.sigmaCur:this.sigmaPot) * 0.9;
    return [clamp(v-s), clamp(v+s)];
  }
  ovrRange(which='cur') {
    const v = which==='cur' ? this.ovr : this.pot;
    const s = (which==='cur'?this.sigmaCur:this.sigmaPot) * 0.50;
    return [clamp(v-s), clamp(v+s)];
  }
  text() {
    const h = this.p.hidden, out = [];
    const we = h.work_ethic;
    out.push(we>=65 ? '훈련 태도가 대단히 성실하다.' : we>=52 ? '성실한 선수로 평가된다.'
      : we<38 ? '훈련 태도에 관한 우려가 있다.' : '훈련 태도는 평범하다.');
    if (h.professionalism>=65) out.push('자기 관리가 뛰어나다.');
    else if (h.professionalism<35) out.push('프로 의식에 물음표가 붙는다.');
    if (h.injury_prone>=63) out.push('부상 이력과 체질이 걱정스럽다.');
    else if (h.injury_prone<=37) out.push('몸이 튼튼하다.');
    if (this.looks<=1) out.push('관찰 기회가 적어 평가의 불확실성이 크다.');
    else if (this.looks>=3) out.push('장기간 추적 관찰한 선수다.');
    return out.join(' ');
  }
}
